src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,386 @@
+        }
+
+        Task IBlockListener.BlockRemovingAsync(Block block, int height, CancellationToken cancellationToken)
+        {
+            return this.watcher.ExecuteAsync(block, height, BlockEventType.Removing, cancellationToken);
+        }
+
+        Task<IEnumerable<Guid>> ITransactionConfirmationWatcherHandler<Guid>.CreateContextsAsync(Transaction tx, CancellationToken cancellationToken)
+        {
+            this.timerLock.EnterReadLock();
+
+            try
+            {
+                if (this.timers.TryGetValue(tx.GetHash(), out var txTimers))
+                {
+                    return Task.FromResult((IEnumerable<Guid>)txTimers.Select(t => t.Key).ToList());
+                }
+            }
+            finally
+            {
+                this.timerLock.ExitReadLock();
+            }
+
+            return Task.FromResult(Enumerable.Empty<Guid>());
+        }
+
+        async Task<bool> IConfirmationWatcherHandler<TransactionWatch<Guid>, Guid>.ConfirmationUpdateAsync(TransactionWatch<Guid> watch, int confirmation, ConfirmationType type, CancellationToken cancellationToken)
+        {
+            switch (type)
+            {
+            case ConfirmationType.Unconfirming:
+                if (confirmation == 1)
+                {
+                    await ResumeTimerAsync(watch.Context);
+                    return false;
+                }
+                break;
+
+            case ConfirmationType.Confirmed:
+                if (confirmation == 1 && !(await StopTimer(watch.TransactionId, watch.Context)))
+                {
+                    return false;
+                }
+
+                this.timerLock.EnterUpgradeableReadLock();
+                try
+                {
+                    var requiredConfirmations = this.timers[watch.TransactionId][watch.Context].Item2;
+
+                    if (confirmation == requiredConfirmations)
+                    {
+                        await ConfirmAsync(watch);
+                        return true;
+                    }
+                }
+                finally
+                {
+                    this.timerLock.ExitUpgradeableReadLock();
+                }
+
+                break;
+
+            default:
+                throw new NotSupportedException($"{nameof(ConfirmationType)} is not supported");
+            }
+
+            return false;
+        }
+
+        Task<IEnumerable<TransactionWatch<Guid>>> IConfirmationWatcherHandler<TransactionWatch<Guid>, Guid>.GetCurrentWatchesAsync(CancellationToken cancellationToken)
+        {
+            var watches = new Collection<TransactionWatch<Guid>>();
+            foreach (var watch in this.watches)
+            {
+                watches.Add(watch.Value);
+            }
+
+            return Task.FromResult(watches.AsEnumerable());
+        }
+
+        Task IWatcherHandler<TransactionWatch<Guid>, Guid>.AddWatchesAsync(IEnumerable<TransactionWatch<Guid>> watches, CancellationToken cancellationToken)
```

Comment:
We need to stop timer and update remaining time for watches that passing to this method here. Here is the steps of handler to get invoke:

1. `CreateContextsAsync` to determine if there is any watch to create by inspecting `Transaction`.
2. `AddWatchesAsync` to store watches that just created from step 1.

When `AddWatchesAsync` get called that mean the passing watches already have one confirmation (but `ConfirmationUpdateAsync` not invoked yet).

Reply:
Good idea, already move timers write to `AddWatchesAsync` and `RemoveWatchAsync` instead.